File: pubsubsystem/topic.go
package pubsubsystem

import (
	"sync"
)

type Topic struct {
	Name        string
	Subscribers map[Subscriber]struct{}
	Messages    []*Message
	mu          sync.RWMutex
}

func NewTopic(name string) *Topic {
	return &Topic{
		Name:        name,
		Subscribers: make(map[Subscriber]struct{}),
		Messages:    []*Message{},
	}
}

func (t *Topic) AddSubscriber(sub Subscriber) {
	t.mu.Lock()
	defer t.mu.Unlock()

	t.Subscribers[sub] = struct{}{}
	if sub.State == Cold {
		for _, message := range t.Messages {
			sub.OnMessage(message)
		}
	}
}

func (t *Topic) RemoveSubscriber(sub Subscriber) {
	t.mu.Lock()
	defer t.mu.Unlock()

	delete(t.Subscribers, sub)
}

func (t *Topic) PublishMsg(message *Message) {
	t.mu.Lock()
	defer t.mu.Unlock()

	t.Messages = append(t.Messages, message)
	for subscriber := range t.Subscribers {
		// subscribe to the message
		subscriber.OnMessage(message)
	}
}

File: pubsubsystem/typescript/topic.ts
import { Subscriber, SubscriberState } from './typescript/subscriber';
import { Message } from './typescript/message';

export class Topic {
    name: string;
    subscribers: Map<Subscriber, boolean>;
    messages: Message[];

    constructor(name: string) {
        this.name = name;
        this.subscribers = new Map<Subscriber, boolean>();
        this.messages = [];
    }

    static newTopic(name: string): Topic {
        return new Topic(name);
    }

    addSubscriber(sub: Subscriber): void {
        this.subscribers.set(sub, true);
        
        if (sub.state === SubscriberState.Cold) {
            for (const message of this.messages) {
                sub.onMessage(message);
            }
        }
    }

    removeSubscriber(sub: Subscriber): void {
        this.subscribers.delete(sub);
    }

    publishMsg(message: Message): void {
        this.messages.push(message);
        
        for (const subscriber of this.subscribers.keys()) {
            // subscribe to the message
            subscriber.onMessage(message);
        }
    }
}